### PR TITLE
fix(model): keep -0 to '-0'

### LINKED ISSUE #11909: Number -0 cannot be correctly rendered in input :value
### Version
2.6.11

### Reproduction link
[https://codepen.io/yuchaosydney/pen/xxRgEPP](https://codepen.io/yuchaosydney/pen/xxRgEPP)

### Steps to reproduce
1. Open reproduction link
2. Compare display value from input tag rendered by Vue and plain input html tag

### What is expected?
Input dom should correctly render value -0

### What is actually happening?
It renders as 0

---
In javascript, Number `-0` is legitimate value, this can be verified by checking `Number(-0) === -0`. Therefore, when input `:value` is number `-0`, input tag should correctly display it, just as a plain input html tag does.

### PR DESCRIPTION
**Pull request by yuwu9145**
fix #11909

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [x] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
In javascript, Number `-0` is legitimate value, this can be verified by checking `Number(-0) === -0`. However, before set a number value to input tag, Vue uses `String(value)` to convert the value to string but `String(-0)` will result in `0` rather than `-0`.

### CODE REVIEW

File: src/platforms/web/runtime/modules/dom-props.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ function updateDOMProps (oldVnode: VNodeWithData, vnode: VNodeWithData) {
       // non-string values will be stringified
       elm._value = cur
       // avoid resetting cursor position when value is the same
-      const strCur = isUndef(cur) ? '' : String(cur)
```
Comment (yuwu9145) [inline src/platforms/web/runtime/modules/dom-props.js:47]: Input tag should be able to display -0. However, String(-0) will result in 0.

File: src/shared/util.js [hunk 2]
```diff
@@ -8,6 +8,10 @@ export function isUndef (v: any): boolean %checks {
   return v === undefined || v === null
 }
 
+export function isNumNegZero (v: any): boolean %checks {
+  return !isNaN(v) && 1/Number(v) === -Infinity
```
Comment (yuwu9145) [inline src/shared/util.js:12]: According [this eslint rule](https://eslint.org/docs/rules/no-compare-neg-zero), `Number(v) === -0` shall not be used. In addition, due to `Object.is` is not supported by IE. Therefore, `1/Number(v) === -Infinity` is used to determine whether `Number(v)` is `-0`

Comment (sirlancelot) [reply src/shared/util.js:12]: In my testing, the `isNaN()` check is unnecessary, as is the `Number(v)` conversion. `return (1 / v) === -Infinity` should do it.

File: src/platforms/web/runtime/modules/dom-props.js [hunk 3]
```diff
@@ -44,7 +44,7 @@ function updateDOMProps (oldVnode: VNodeWithData, vnode: VNodeWithData) {
...
+      const strCur = isNumNegZero(cur) ? '-0' : isUndef(cur) ? '' : String(cur)
```
Comment (sirlancelot) [inline src/platforms/web/runtime/modules/dom-props.js:47]: I think in the interest of least impactful, this collection of ternaries should keep the `isUndef()` at the top like so:

```js
const strCur = isUndef(cur) ? '' : isNumNegZero(cur) ? '-0' : String(cur)
```

This keeps the case of `cur` being empty as the fastest/shortest path.

**Comment by Justineo** [comment]
Hi, can you add some test cases to reflect these changes? (which only affects `v-model.number` while `v-model`'s original behavior is kept, according to the discussions in #11909)

**Comment by yuwu9145** [comment]
> Hi, can you add some test cases to reflect these changes? (which only affects `v-model.number` while `v-model`'s original behavior is kept, according to the discussions in #11909)

HI @Justineo , yes, I am happy to add test case. I will also consider and include potential changes suggested by @sirlancelot

**Comment by posva** [comment]
As said, I don't think it should modify don props, which is perf critical. It should only apply to vmodel number

**Comment by yuwu9145** [comment]
> As said, I don't think it should modify don props, which is perf critical. It should only apply to vmodel number

@posva What about making `v-model` & `v-model.number` output `'-0': string` only for `-0: number`? Then pass `'-0': string` to `dom-props`? So it does not have to make any change in `dom-props`.

**Comment by yuwu9145** [comment]
After thinking through all concerns & reminding this is a very rare edge case. Probably let `v-model.number` handle `-0:number` is enough, so the only change has been made is:

`-0: number -> v-model.number -> '-0': string`

Please advice if more test case or changes are required

File: src/shared/util.js [hunk 4]
```diff
@@ -92,10 +96,11 @@ export function toString (val: any): string {
 
 /**
  * Convert an input value to a number for persistence.
+ * Check for special case of signed zero.
  * If the conversion fails, return original string.
  */
 export function toNumber (val: string): number | string {
-  const n = parseFloat(val)
+  const n = isNumNegZero(val) ? '-0' : parseFloat(val)
```
Comment (posva) [inline src/shared/util.js:103]: let's inline the check and remove the function

```suggestion
  const n = 1 / val === -Infinity ? '-0' : parseFloat(val)
```

**Review by posva — CHANGES_REQUESTED** [review]
I don't think this is still good as, I will say it again, it should be an edge case of vmodel. It's also important that v-model.number sets the value as `-0` and not `'-0'` when typing `-0`. Could you add both unit tests?

**Comment by yuwu9145** [comment]
@posva the problem is if `v-model.number` outputs `-0: number`, dom-props will do `String(-0)` and result in '0: string' still. I really think a fix should be done in dom-props as well. But if it has a performance concern, then it has to make a decision between **performance** and **"-0 support"**

**Comment by posva** [comment]
It's not just perf, it's also being consistent with js APIs and not breaking the existing v-model.number when the input has `-0` in it. To be fair, this is still just a small patch, it won't allow people to have a number input that handles the other mentioned edge cases, so I think it's better to just drop the idea. Thanks a lot for your efforts!

**Comment by yuwu9145** [comment]
@posva no worries. It's good know the concern
